src/transfer.cpp

```diff
@@ -398,29 +398,36 @@ void Transfer::failed(error e, DBTableTransactionCommitter& committer, dstime ti
             client->app->transfer_failed(this, e, timeleft);
             ++client->performanceStats.transferTempErrors;
         }
-        else
+        else    // already dispatched (slot != null)
         {
             bool allForeignTargets = true;
-            for (auto &file : files)
+            for (const auto &file : files)
             {
                 if (client->isPrivateNode(file->h))
                 {
                     allForeignTargets = false;
                     break;
                 }
             }
-
-            /* If all targets are foreign and there's not a bandwidth overquota, transfer must fail.
-             * Otherwise we need to activate overquota.
-             */
-            if (!timeleft && allForeignTargets)
+            if (!timeleft && allForeignTargets)   // storage overquota, but targets are foreign
             {
                 client->app->transfer_failed(this, e);
             }
-            else
+            else    // bandwidth overquota OR storage overquota for, at least, a private target
             {
                 bt.backoff(timeleft ? timeleft : NEVER);
-                client->activateoverquota(timeleft);
+                if (client->ststatus == STORAGE_RED && !timeleft)   // already in storage overquota, notify transfer error
+                {
+                    state = TRANSFERSTATE_RETRYING;
+                    slot->retrybt.backoff(NEVER);
+                    slot->retrying = true;
+                    client->app->transfer_failed(this, API_EOVERQUOTA, 0);
+                    ++client->performanceStats.transferTempErrors;
+                }
```

Comment:
In order for client->ststatus to be in STORAGE_RED, shouldn't void `MegaClient::activateoverquota` have been called already, which would have iterated through all the transfers and put them in retrying state, and called transfer_failed for the client?   I think we might be doubling up on the notifications here.  Or at least doubling up on notifications for that case.  Perhaps this is intended for a different case where the transfer was started after we are already in overquota state, but I think we should make sure that we don't impact the transition to overquota?  thanks

Reply:
That's exactly the case: the `activateoverquota()`, in case there's a transition to storage overquota (not when it's already overquota, when it does nothing), notifies the retrying state for all existing transfers. However, this case is a reported bug: once the account is in overquota, new transfers are not notified via `onTransferTemporaryError(OVERQUOTA)`, so the app doesn't show the error in the UI, but the transfer does not progress :)

Reply:
thanks, I see - but for that initial transition to overquota, say we have 30 transfers in progress, the first one of those to discover the account is overquota will call this functoin, activateoverquota() will notify the app for all of them, but then the other 29 will also call this function, I think, and follow the new code path - is it possible for us to differentiate those caess?  thanks

Reply:
In this situation the first failed transfer will call activate overquota and will set storage status as RED, so the other failed transfers will set it's state as retrying, adjust it's backoff and notify the app a temporary error but no call activate overquota again.